Mathlib/Analysis/Normed/MulAction.lean

```diff
@@ -121,4 +119,25 @@ theorem nndist_smul₀ (s : α) (x y : β) : nndist (s • x) (s • y) = ‖s
 theorem edist_smul₀ (s : α) (x y : β) : edist (s • x) (s • y) = ‖s‖₊ • edist x y := by
   simp only [edist_nndist, nndist_smul₀, ENNReal.coe_mul, ENNReal.smul_def, smul_eq_mul]
 
-end NormedDivisionRingModule
+instance NormSMulClass.toBoundedSMul : IsBoundedSMul α β where
+  dist_smul_pair' x y₁ y₂ := by simp [dist_smul₀, dist_zero_right]
+  dist_pair_smul' x₁ x₂ y := by simpa [dist_eq_norm, sub_smul] using (norm_smul (x₁ - x₂) y).le
+
+end NormSMulClassModule
+
+section NormedDivisionRing
+
+variable [NormedDivisionRing α] [SeminormedAddGroup β]
+variable [MulActionWithZero α β] [IsBoundedSMul α β]
+
+instance NormedDivisionRing.toNormSMulClass : NormSMulClass α β where
```

Comment:
I think that these new instances should have priority 100. Also, please add "structural" instances for action on a product and on a pi type.

Reply:
This instance looks like a bad idea to me, as it forms a loop with the one above.

Reply:
I think instead the instance should be from `NormedSpace`

Reply:
This won't work for normed division rings.

Reply:
What do you mean?

Reply:
The definition of NormedSpace requires commutativity (a NormedField rather than a NormedDivisionRing).

Reply:
I think what I'm suggesting is:
```
instance [NormedModule A B] : NormSMulClass A B
instance [NormMulClass A] : NormSMulClass A A
instance [NormMulClass A] : NormSMulClass (MulOpposite A) A
```
which should cover all the cases we care about?

Reply:
This would not cover modules over a non-commutative normed division ring (e.g. quaternions) and there are certainly people who care about those. Incidentally, we don't have a `NormedModule` typeclass, this is usually spelled long-hand as `[Module A B] [NormedAddCommGroup A B] [IsBoundedSMul A B]`.

Reply:
Sorry, that was a typo. I claim that if people are talking about abstract normed modules over the quaternions, they should write
```lean
variable [Module H B] [NormedAddCommGroup B] [NormSMulClass H B]
```
instead of
```lean
variable [Module H B] [NormedAddCommGroup B] [IsBoundedSMul H B]
```
and then there is no need for this loop-forming instance.


For any concrete `B`, hopefully the necessary `IsBoundedSMul` would already be found through another route.

Reply:
> Sorry, that was a typo. I claim that if people are talking about abstract normed modules over the quaternions, they should write [...]

But that would be forcing these people to prove more, so this is a (small) loss of functionality. I guess we could keep the existing argument for normed division rings as a lemma rather than an instance, and then have an instance for `NormedSpace` which invokes that lemma.

Reply:
I implemented the idea discussed above. I'm not 100% happy with it, because it means we have 3 different "standard spellings" for declaring a normed module depending on the kind of ring concerned -- over a general normed ring it's `IsBoundedSMul`, over a non-commutative division ring it's `NormMulClass`, and over a normed field it's `NormedSpace`. I wonder if we could change the constructor for `NormedSpace` to accept non-commutative normed division rings.

Reply:
Thanks! Could you fix the build errors so that we can benchmark it?

Long ago I considered redefining `NormedSpace K V` to be the type of K-bimodules over V, roughly as
```lean
class NormedSpace K V extends Module K V, Module (MulOpposite K) V, IsCentralScalar K V, NormSMulClass K V
```
which would be a very weak vote against

> I wonder if we could change the constructor for `NormedSpace` to accept non-commutative normed division rings.

Reply:
> Thanks! Could you fix the build errors so that we can benchmark it?

I'm trying! The problem is that there are 1000 files depending on `Normed/MulAction`, so it takes a very long time to rebuild; and each time I run the build it reveals new errors. It's very hard to search the code for examples of `IsBoundedSMul` over division rings.

Reply:
Re: 3 ways to talk about the same thing. Given what's happening to the algebra+order hierarchies, I suspect that we're going to drop `NormedSpace` in favor of the class you're adding now.